pep-0xxx.rst

```diff
@@ -0,0 +1,727 @@
+            ``TLSWrappedSocket`` object. ``socket`` must be a ``SOCK_STREAM``
+            socket: all other socket types are unsupported.
+
+            The returned SSL socket is tied to the context, its settings and
+            certificates.
+
+            The parameter ``auto_handshake`` specifies whether to do the SSL
+            handshake automatically after doing a ``socket.connect()``, or
+            whether the application program will call it explicitly, by
+            invoking the ``TLSWrappedSocket.do_handshake()`` method. Calling
+            ``TLSWrappedSocket.do_handshake()`` explicitly gives the program
+            control over the blocking behavior of the socket I/O involved in
+            the handshake.
+            """
+
+        @abstractmethod
+        def wrap_buffers(self, incoming: Any, outgoing: Any) -> TLSWrappedBuffer:
+            """
+            Wrap a pair of buffer objects (``incoming`` and ``outgoing``) to
+            create an in-memory stream for TLS. The SSL routines will read data
+            from ``incoming`` and decrypt it, and write encrypted data to
+            ``outgoing``.
+            """
+
+
+Socket
+~~~~~~
+
+The socket-wrapper ABC will be defined by the ``TLSWrappedSocket`` ABC, which
+has the following definition::
+
+    class TLSWrappedSocket(metaclass=ABCMeta):
+        # The various socket methods all must be implemented. Their definitions
+        # have been elided from this class defintion in the PEP because they
+        # aren't instructive.
+        @abstractmethod
+        def do_handshake(self) -> None:
+            """
+            Performs the TLS handshake. Also performs certificate validation
+            and hostname verification.
+            """
+
+        @abstractmethod
+        def cipher(self) -> Optional[Cipher]:
+            """
+            Returns the Cipher entry for the cipher that has been negotiated on
+            the connection. If no connection has been negotiated, returns
+            ``None``.
+            """
+
+        @abstractmethod
+        def negotiated_protocol(self) -> Optional[NextProtocol]:
+            """
+            Returns the protocol that was selected during the TLS handshake.
+            This selection may have been made using ALPN, NPN, or some future
+            negotiation mechanism.
+
+            If ``Context.set_inner_protocols()`` was not called, if the other
+            party does not support protocol negotiation, if this socket does
+            not support any of the peer's proposed protocols, or if the
+            handshake has not happened yet, ``None`` is returned.
+            """
+
+        @property
+        @abstractmethod
+        def context(self) -> Context:
+            """
+            The ``Context`` object this socket is tied to.
+            """
+
+        @context.setter
+        @abstractmethod
+        def context(self, value: Context) -> None:
```

Comment:
The context-swapping API? Yeah, I can get behind removing that. I guess the question becomes: how do you request a change to the TLS settings based on the SNI parameter? Should you be able to do that?